### PR TITLE
Trigger server startup

### PR DESCRIPTION
**Pull request by zalokhan**
Trigger server startup.
Sequencer and Logunit server calls are blocked until Management Servers signals startup.

### CODE REVIEW

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **d0a909e**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #762 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/7RIN253RY6uJp5D2VwQk7nLRGpOZSvXs)

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java [hunk 1]
```diff
@@ -209,6 +215,9 @@ private void read(CorfuPayloadMsg<ReadRequest> msg, ChannelHandlerContext ctx, I
     private void fillHole(CorfuPayloadMsg<TrimRequest> msg, ChannelHandlerContext ctx,
                           IServerRouter r,
                           boolean isMetricsEnabled) {
+        if (!checkServerStart(msg, ctx, r, serverContext)) {
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: Factor out this to a method? void isRead(params)?

File: infrastructure/src/main/java/org/corfudb/infrastructure/AbstractServer.java [hunk 2]
```diff
@@ -26,6 +27,16 @@ public AbstractServer() {
      */
     public abstract CorfuMsgHandler getHandler();
 
+    boolean checkServerStart(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r,
+                                     ServerContext serverContext) {
+        if (serverContext.isStart()) {
+            return true;
+        } else {
+            r.sendResponse(ctx, msg, CorfuMsgType.NACK.msg());
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/AbstractServer.java]: maybe call this NOT_READY ?

**Comment by zalokhan** [comment]
@Maithem I've made some changes to the message handling. Let me know if this is better or still needs changes.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **83c9390**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #762 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/I51Fun1fa0GKZ9kAXdDb6Pn7iPtJSK3o)

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java [hunk 3]
```diff
@@ -102,11 +102,19 @@
 
     private static final String metricsPrefix = "corfu.server.logunit.";
 
+    private final ServerContext serverContext;
+
+    @Override
+    public boolean isServerReady() {
+        return serverContext.isStart();
```
Comment (medhavidhawan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: This flag should be more like ready and not start but this is just a nit.

Comment (zalokhan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: I'll make the change.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java [hunk 4]
```diff
@@ -158,11 +164,12 @@ public ManagementServer(ServerContext serverContext) {
 
     private void recover() {
         try {
-            failureHandlerDispatcher.recoverCluster((Layout) latestLayout.clone(),
+            failureHandlerDispatcher.recoverCluster(serverContext, (Layout) latestLayout.clone(),
                     getCorfuRuntime());
             safeUpdateLayout(corfuRuntime.getLayoutView().getLayout());
         } catch (CloneNotSupportedException e) {
             log.error("Failure Handler could not clone layout: {}", e);
+            log.error("Node will remain blocked until recovered successfully.");
```
Comment (no2chem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: Add Recover: to the log message

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/CorfuMsgType.java [hunk 5]
```diff
@@ -30,6 +30,7 @@
     NACK(6, TypeToken.of(CorfuMsg.class)),
     VERSION_REQUEST(7, TypeToken.of(CorfuMsg.class), true),
     VERSION_RESPONSE(8, new TypeToken<JSONPayloadMsg<VersionInfo>>() {}, true),
+    NOT_READY(9, TypeToken.of(CorfuMsg.class), true),
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/protocols/wireprotocol/CorfuMsgType.java]: Ok. I think at some point we'll want to merge all the XX_NO_BOOTSTRAP messages to this one.

File: runtime/src/main/java/org/corfudb/runtime/view/AbstractView.java [hunk 6]
```diff
@@ -83,6 +84,15 @@ public Layout getCurrentLayout() {
                     } catch (InterruptedException ie) {
                         log.warn("Interrupted Exception in layout helper.", ie);
                     }
+                } else if (re instanceof ServerNotReadyException) {
+                    log.warn("Server still not ready. Waiting for server to start "
+                            + "accepting requests.");
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/view/AbstractView.java]: If the server is not ready, why do you invalidate the layout ?

Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/runtime/view/AbstractView.java]: You are right, invalidation should be triggered automatically after a seal with a WrongEpochException.
I'll remove this.

File: infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java (hunk 4 above)
Comment (zalokhan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java]: Will add this.

**Comment by Maithem** [comment]
Nvm, discard my last post. It seems you addressed it already.

File: infrastructure/src/main/java/org/corfudb/infrastructure/LayoutServer.java [hunk 7]
```diff
@@ -81,6 +81,11 @@
     private CorfuMsgHandler handler = new CorfuMsgHandler()
             .generateHandlers(MethodHandles.lookup(), this);
 
+    @Override
+    public boolean isServerReady() {
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LayoutServer.java]: do you need to override this? the base server already returns true?

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java [hunk 8]
```diff
@@ -102,11 +101,19 @@
 
     private static final String metricsPrefix = "corfu.server.logunit.";
 
+    private final ServerContext serverContext;
+
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: Should this be part of the server context, or a variable in the base server?

File: runtime/src/main/java/org/corfudb/runtime/exceptions/ServerNotReadyException.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package org.corfudb.runtime.exceptions;
+
+/**
+ * Created by zlokhandwala on 6/26/17.
+ */
+public class ServerNotReadyException extends RuntimeException {
+
+    public ServerNotReadyException() {
+        super("Server not ready to accept messages.");
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/exceptions/ServerNotReadyException.java]: Exception already indicates that?

File: infrastructure/src/main/java/org/corfudb/infrastructure/FailureHandlerDispatcher.java [hunk 10]
```diff
@@ -32,14 +32,18 @@
...
+    public boolean recoverCluster(ServerContext serverContext, Layout recoveryLayout,
+                                  CorfuRuntime corfuRuntime) {
 
         try {
             // Seals and increments the epoch.
             recoveryLayout.setRuntime(corfuRuntime);
             sealEpoch(recoveryLayout);
             log.info("After seal: {}", recoveryLayout.asJSONString());
 
+            // Mark cluster as ready after sealing.
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/FailureHandlerDispatcher.java]: Why do you need this? if other servers are also on the management server wouldn't you send a message to yourself and change the ready status via the handler?

Comment (zalokhan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/FailureHandlerDispatcher.java]: Sorry, this comment should be "Mark node as ready after sealing"
The status is just communicated via the shared context between the different components on the same node.

File: runtime/src/main/java/org/corfudb/runtime/exceptions/ServerNotReadyException.java (hunk 9 above)
Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/runtime/exceptions/ServerNotReadyException.java]: Yes this is redundant.

File: infrastructure/src/main/java/org/corfudb/infrastructure/LayoutServer.java (hunk 7 above)
Comment (zalokhan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LayoutServer.java]: All servers have different implementations.
Base, Layout and Management Servers are by default in the ready state so their implementations return true directly.
The Logunit and the Sequencer servers start in a non-ready state and have to wait for the isReady flag to be true.

**Comment by zalokhan** [comment]
The flag is present only in the serverContext which is shared among all the components: Base, Layout, Sequencer, Logunit and Management servers.

The `abstract boolean isServerReady()` in the AbstractServer is an abstract method.
The implementation if this method in Sequencer and Logunit server, checks the serverContext and returns true or false depending on the `isReady` flag.
Base, Layout and Management are started in the ready state and do not require explicit triggering.

I can rename `abstract boolean isServerReady()` -> `abstract boolean checkIfServerReady()`
Does this simplify?

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java (hunk 8 above)
Comment (zalokhan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: The base server is not shared among all component servers.
Using the serverContext management server can signal all servers to start accepting requests.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **93dc1f5**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #762 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/jb4mfDYhg1WxubhkgLbWarqmtJVeETBB)

**Pull request merged**